Remove tmpdir from tests

matthewhughes934 opened the pull request:
<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [ ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->

Some more cleanup of `tmpdir` from the test suite

* Remove some unused 'tmpdir's
* Replace some 'tmpdirs' with 'tmp_path' in makepyfile calls
* Rename variables 'tmpdir'->'tmp_path'

    Rename this variables reflecting the migrations made e.g. with 3bcd316f0
    and ed658d682

Following this, the remaining usages (trying to ignore `from _pytest.tmpdir import ...`) look like:

```
$ git grep --word-regexp --perl-regexp --count '(?<!from _pytest\.)tmpdir' -- testing/
testing/example_scripts/tmpdir/tmpdir_fixture.py:3
testing/python/fixtures.py:9
testing/test_tmpdir.py:15
```

I think there are some more usages in `testing/python/fixtures.py` that could possibly be replace, but the others all look to be explicitly testing `tmpdir`

RonnyPfannschmidt commented on testing/python/collect.py:
```diff
@@ -274,7 +274,7 @@ def test_function_as_object_instance_ignored(self, pytester: Pytester) -> None:
             """
             class A(object):
-                def __call__(self, tmpdir):
+                def __call__(self):
```
This intentionally has a fixture name for the signature

matthewhughes934 replied:
> This intentionally has a fixture name for the signature

Ah, in that case would you like it renamed `tmpdir->tmp_path` (I've been proceeding under the assumption that even for non-functional use-cases there's value in renaming just to make things clearer).

Out of curiosity, what's the intent in having a fixture name in the signature? I couldn't see anything explicitly wanting it in the test (though admittedly I didn't dive too deep into the surrounding context)

The-Compiler replied:
+1 for renaming it to `tmp_path` for consistency.

As for your question - since the test is named `test_function_as_object_instance_ignored` (and also looking at 24db492f53a7391c2bbc660adb0c44f072dfca58), the test seems to ensure that objects named `test_x` aren't collected. Part of that collection is also filling fixture values, so we'd want to make sure that this doesn't break.

matthewhughes934 replied:
> As for your question - since the test is named test_function_as_object_instance_ignored (and also looking at 24db492), the test seems to ensure that objects named test_x aren't collected. Part of that collection is also filling fixture values, so we'd want to make sure that this doesn't break.

Thanks, do you think there would be any value in being more explicit about this? e.g.

```python
            import pytest

            @pytest.fixture
            def never_filled():
                assert False, "Fixture should never be filled!"

            class A(object):
                def __call__(self, never_filled):
                    0/0

            test_a = A()
```

matthewhughes934 replied:
:+1: restored + renamed c9bb4c4

The-Compiler replied:
I don't really have an opinion on making it more explicit vs. the way it is currently. Surely wouldn't hurt, but I don't mind much either way.

The-Compiler commented on testing/python/integration.py:
```diff
@@ -234,7 +234,7 @@ def mock_basename(path):
             @mock.patch("os.path.normpath")
             @mock.patch("os.path.basename", new=mock_basename)
-            def test_someting(normpath, abspath, tmpdir):
+            def test_someting(normpath, abspath):
```
Looking at dba2a8bc647c73eb9ecf21de18bf56e45a04f115 this is deliberate as well - to ensure that `mock.patch` and fixtures work together well.

matthewhughes934 replied:
> Looking at dba2a8b this is deliberate as well - to ensure that mock.patch and fixtures work together well.

:+1: restored + renamed c9bb4c4

bluetech commented on testing/test_collection.py:
```diff
@@ -127,16 +127,16 @@ def test_foo():
+        ensure_file(tmp_path / ".whatever" / "test_notfound.py")
+        ensure_file(tmp_path / ".bzr" / "test_notfound.py")
+        ensure_file(tmp_path / "normal" / "test_found.py")
+        for x in Path(str(tmp_path)).rglob("test_*.py"):
```
Can remove the conversion to Path now:

```suggestion
        for x in tmp_path.rglob("test_*.py"):
```

bluetech commented on testing/test_collection.py:
```diff
@@ -628,9 +630,9 @@ def test_method(self):
-        with tmpdir.cwd():
+        tmp_path = pytester.path
+        x = ensure_file(tmp_path / "x.py")
+        with tmp_path.cwd():
```
This doesn't do the same thing. In `py.path`, the `cwd()` context manager does a `chdir()` to the path on enter and restores to previous value on exit. In `pathlib`, the context manager doesn't do anything (I have no idea why it even exists...).

The replacement is something like

```py
with MonkeyPatch.context() as mp:
    mp.chdir(tmp_path)
```

Although now that I think of it, `pytester` already does a `chdir` into its path, so the `with` is actually unnecessary and you can just remove it :smile:

bluetech reviewed (commented):
I left 2 comments, the rest is good!

Also adding a reference: #8204.

matthewhughes934 commented:
> I left 2 comments, the rest is good!

Thanks, addressed both with 514f8e0

> Also adding a reference: #8204.

Ah, I meant to do this once I moved the PR out of draft :facepalm: thanks again
